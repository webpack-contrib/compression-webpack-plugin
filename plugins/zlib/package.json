{
  "name": "compression-webpack-plugin-zlib",
  "version": "1.0.0",
  "author": "Tobias Koppers @sokra",
  "maintainers": [
    "Jason Palmer @jsonp"
  ],
  "description": "Prepare compressed versions of assets to serve them with Content-Encoding",
  "dependencies": {
    "async": "0.2.x",
    "webpack-sources": "^0.1.0"
  },
  "peerDependencies": {
    "webpack": ">=0.11 <2.0.0 || >=2.1.0-beta.13 <3.0.0"
  },
  "devDependencies": {
    "mocha": "^2.3.2"
  },
  "scripts": {
    "test": "mocha --reporter min 'test/**/*test*.js'"
  },
  "files": [
    "index.js"
  ],
  "homepage": "http://github.com/webpack/compression-webpack-plugin",
  "repository": {
    "type": "git",
    "url": "http://github.com/webpack/compression-webpack-plugin.git"
  },
  "licenses": [
    {
      "type": "MIT",
      "url": "http://www.opensource.org/licenses/mit-license.php"
    }
  ]
}
