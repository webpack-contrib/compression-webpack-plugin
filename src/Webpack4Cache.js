import os from 'os';

import cacache from 'cacache';
import findCacheDir from 'find-cache-dir';
import serialize from 'serialize-javascript';

export default class Webpack4Cache {
  constructor(compilation, options, weakCache) {
    this.cache =
      options.cache === true
        ? Webpack4Cache.getCacheDirectory()
        : options.cache;
    this.weakCache = weakCache;
  }

  static getCacheDirectory() {
    return findCacheDir({ name: 'compression-webpack-plugin' }) || os.tmpdir();
  }

  async get(task, sources) {
    const weakOutput = this.weakCache.get(task.assetSource);

    if (weakOutput) {
      return weakOutput;
    }

    if (!this.cache) {
      // eslint-disable-next-line no-undefined
      return undefined;
    }

    // eslint-disable-next-line no-param-reassign
    task.cacheIdent = task.cacheIdent || serialize(task.cacheKeys);

    let cachedResult;

    try {
      cachedResult = await cacache.get(this.cache, task.cacheIdent);
    } catch (ignoreError) {
      // eslint-disable-next-line no-undefined
      return undefined;
    }

    return new sources.RawSource(
      Buffer.from(JSON.parse(cachedResult.data).data)
    );
  }

  async store(task) {
    if (!this.weakCache.has(task.assetSource)) {
      this.weakCache.set(task.assetSource, task.output);
    }

    if (!this.cache) {
      // eslint-disable-next-line no-undefined
      return undefined;
    }

    const { cacheIdent, output } = task;

    return cacache.put(this.cache, cacheIdent, JSON.stringify(output.source()));
  }
}
