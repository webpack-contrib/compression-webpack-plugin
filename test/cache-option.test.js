import path from 'path';
import zlib from 'zlib';

import cacache from 'cacache';
import findCacheDir from 'find-cache-dir';
import del from 'del';

import CompressionPlugin from '../src/index';
import Webpack4Cache from '../src/Webpack4Cache';

import {
  compile,
  getAssetsNameAndSize,
  getCompiler,
  getErrors,
  getWarnings,
} from './helpers/index';

const cacheDir = findCacheDir({ name: 'compression-webpack-plugin' });
const otherCacheDir = findCacheDir({ name: 'other-cache-directory' });
const uniqueDirectory = findCacheDir({ name: 'unique-cache-directory' });

if (getCompiler.isWebpack4()) {
  describe('"cache" option', () => {
    beforeEach(() => {
      return Promise.all([
        cacache.rm.all(cacheDir),
        cacache.rm.all(otherCacheDir),
        cacache.rm.all(uniqueDirectory),
      ]);
    });

    afterEach(() =>
      Promise.all([cacache.rm.all(cacheDir), cacache.rm.all(otherCacheDir)])
    );

    it('matches snapshot for `false` value ({Boolean})', async () => {
      const compiler = getCompiler('./entry.js');

      new CompressionPlugin({ cache: false, minRatio: 1 }).apply(compiler);

      cacache.get = jest.fn(cacache.get);
      cacache.put = jest.fn(cacache.put);

      const stats = await compile(compiler);

      expect(getAssetsNameAndSize(stats)).toMatchSnapshot('assets');
      expect(getWarnings(stats)).toMatchSnapshot('errors');
      expect(getErrors(stats)).toMatchSnapshot('warnings');

      // Cache disabled so we don't run `get` or `put`
      expect(cacache.get.mock.calls.length).toBe(0);
      expect(cacache.put.mock.calls.length).toBe(0);

      const cacheEntriesList = await cacache.ls(cacheDir);
      const cacheKeys = Object.keys(cacheEntriesList);

      expect(cacheKeys.length).toBe(0);
    });

    it('matches snapshot for `true` value ({Boolean})', async () => {
      const beforeCacheCompiler = getCompiler('./entry.js');

      new CompressionPlugin({ cache: true, minRatio: 1 }).apply(
        beforeCacheCompiler
      );

      cacache.get = jest.fn(cacache.get);
      cacache.put = jest.fn(cacache.put);

      const getCacheDirectorySpy = jest
        .spyOn(Webpack4Cache, 'getCacheDirectory')
        .mockImplementation(() => {
          return uniqueDirectory;
        });

      const stats = await compile(beforeCacheCompiler);

      expect(getAssetsNameAndSize(stats)).toMatchSnapshot('assets');
      expect(getWarnings(stats)).toMatchSnapshot('errors');
      expect(getErrors(stats)).toMatchSnapshot('warnings');

      const countAssets = Object.keys(stats.compilation.assets).length;

      // Try to found cached files, but we don't have their in cache
      expect(cacache.get.mock.calls.length).toBe(countAssets / 2);
      // Put files in cache
      expect(cacache.put.mock.calls.length).toBe(countAssets / 2);

      const cacheEntriesList = await cacache.ls(uniqueDirectory);

      const cacheKeys = Object.keys(cacheEntriesList);

      // Make sure that we cached files
      expect(cacheKeys.length).toBe(countAssets / 2);

      cacheKeys.forEach((cacheEntry) => {
        // eslint-disable-next-line no-new-func
        const cacheEntryOptions = new Function(
          `'use strict'\nreturn ${cacheEntry}`
        )();
        const basename = path.basename(cacheEntryOptions.filename);

        expect([basename, cacheEntryOptions.contentHash]).toMatchSnapshot(
          basename
        );
      });

      cacache.get.mockClear();
      cacache.put.mockClear();

      const afterCacheCompiler = getCompiler('./entry.js');

      new CompressionPlugin({ cache: true, minRatio: 1 }).apply(
        afterCacheCompiler
      );

      const newStats = await compile(afterCacheCompiler);

      expect(getAssetsNameAndSize(newStats)).toMatchSnapshot('assets');
      expect(getWarnings(newStats)).toMatchSnapshot('errors');
      expect(getErrors(newStats)).toMatchSnapshot('warnings');

      const newCountAssets = Object.keys(newStats.compilation.assets).length;

      // Now we have cached files so we get their and don't put
      expect(cacache.get.mock.calls.length).toBe(newCountAssets / 2);
      expect(cacache.put.mock.calls.length).toBe(0);

      getCacheDirectorySpy.mockRestore();
    });

    it('matches snapshot for `other-cache-directory` value ({String})', async () => {
      const beforeCacheCompiler = getCompiler('./entry.js');

      new CompressionPlugin({ cache: otherCacheDir, minRatio: 1 }).apply(
        beforeCacheCompiler
      );

      cacache.get = jest.fn(cacache.get);
      cacache.put = jest.fn(cacache.put);

      const stats = await compile(beforeCacheCompiler);

      expect(getAssetsNameAndSize(stats)).toMatchSnapshot('assets');
      expect(getWarnings(stats)).toMatchSnapshot('errors');
      expect(getErrors(stats)).toMatchSnapshot('warnings');

      const countAssets = Object.keys(stats.compilation.assets).length;

      // Try to found cached files, but we don't have their in cache
      expect(cacache.get.mock.calls.length).toBe(countAssets / 2);
      // Put files in cache
      expect(cacache.put.mock.calls.length).toBe(countAssets / 2);

      const cacheEntriesList = await cacache.ls(otherCacheDir);
      const cacheKeys = Object.keys(cacheEntriesList);

      // Make sure that we cached files
      expect(cacheKeys.length).toBe(countAssets / 2);

      cacheKeys.forEach((cacheEntry) => {
        // eslint-disable-next-line no-new-func
        const cacheEntryOptions = new Function(
          `'use strict'\nreturn ${cacheEntry}`
        )();
        const basename = path.basename(cacheEntryOptions.filename);

        expect([basename, cacheEntryOptions.contentHash]).toMatchSnapshot(
          basename
        );
      });

      cacache.get.mockClear();
      cacache.put.mockClear();

      const afterCacheCompiler = getCompiler('./entry.js');

      new CompressionPlugin({ cache: otherCacheDir, minRatio: 1 }).apply(
        afterCacheCompiler
      );

      const newStats = await compile(afterCacheCompiler);

      expect(getAssetsNameAndSize(newStats)).toMatchSnapshot('assets');
      expect(getWarnings(newStats)).toMatchSnapshot('errors');
      expect(getErrors(newStats)).toMatchSnapshot('warnings');

      const newCountAssets = Object.keys(newStats.compilation.assets).length;

      // Now we have cached files so we get their and don't put
      expect(cacache.get.mock.calls.length).toBe(newCountAssets / 2);
      expect(cacache.put.mock.calls.length).toBe(0);
    });
  });
} else {
  describe('"cache" option', () => {
    beforeEach(() => {
      return del([path.resolve('node_modules/.cache/webpack')]);
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    it('should work when `cache` is `false`', async () => {
      const gzipSpy = jest.spyOn(zlib, 'gzip');
      const beforeCacheCompiler = getCompiler(
        './entry.js',
        {},
        { cache: false }
      );

      new CompressionPlugin().apply(beforeCacheCompiler);

      const stats = await compile(beforeCacheCompiler);

      expect(gzipSpy).toHaveBeenCalledTimes(4);
      expect(getAssetsNameAndSize(stats)).toMatchSnapshot('assets');
      expect(getWarnings(stats)).toMatchSnapshot('errors');
      expect(getErrors(stats)).toMatchSnapshot('warnings');

      gzipSpy.mockClear();

      await new Promise((resolve) => {
        beforeCacheCompiler.close(async () => {
          const afterCacheCompiler = getCompiler(
            './entry.js',
            {},
            { cache: false }
          );

          new CompressionPlugin().apply(afterCacheCompiler);

          const newStats = await compile(afterCacheCompiler);

          expect(gzipSpy).toHaveBeenCalledTimes(4);
          expect(getAssetsNameAndSize(newStats)).toMatchSnapshot('assets');
          expect(getWarnings(newStats)).toMatchSnapshot('errors');
          expect(getErrors(newStats)).toMatchSnapshot('warnings');

          gzipSpy.mockRestore();

          resolve();
        });
      });
    });

    it('should work when `cache` is `memory`', async () => {
      const gzipSpy = jest.spyOn(zlib, 'gzip');
      const compiler = getCompiler(
        './entry.js',
        {},
        { cache: { type: 'memory' } }
      );

      new CompressionPlugin().apply(compiler);

      const stats = await compile(compiler);

      expect(gzipSpy).toHaveBeenCalledTimes(4);
      expect(getAssetsNameAndSize(stats)).toMatchSnapshot('assets');
      expect(getWarnings(stats)).toMatchSnapshot('errors');
      expect(getErrors(stats)).toMatchSnapshot('warnings');

      gzipSpy.mockClear();

      await new Promise((resolve) => {
        compiler.close(async () => {
          const newStats = await compile(compiler);

          expect(gzipSpy).toHaveBeenCalledTimes(0);
          expect(getAssetsNameAndSize(newStats)).toMatchSnapshot('assets');
          expect(getWarnings(newStats)).toMatchSnapshot('errors');
          expect(getErrors(newStats)).toMatchSnapshot('warnings');

          gzipSpy.mockRestore();

          resolve();
        });
      });
    });

    it('should work when `cache` is `filesystem`', async () => {
      const gzipSpy = jest.spyOn(zlib, 'gzip');
      const beforeCacheCompiler = getCompiler(
        './entry.js',
        {},
        { cache: { type: 'filesystem' } }
      );

      new CompressionPlugin().apply(beforeCacheCompiler);

      const stats = await compile(beforeCacheCompiler);

      expect(gzipSpy).toHaveBeenCalledTimes(4);
      expect(getAssetsNameAndSize(stats)).toMatchSnapshot('assets');
      expect(getWarnings(stats)).toMatchSnapshot('errors');
      expect(getErrors(stats)).toMatchSnapshot('warnings');

      gzipSpy.mockClear();

      await new Promise((resolve) => {
        beforeCacheCompiler.close(async () => {
          const afterCacheCompiler = getCompiler(
            './entry.js',
            {},
            { cache: { type: 'filesystem' } }
          );

          new CompressionPlugin().apply(afterCacheCompiler);

          const newStats = await compile(afterCacheCompiler);

          expect(gzipSpy).toHaveBeenCalledTimes(0);
          expect(getAssetsNameAndSize(newStats)).toMatchSnapshot('assets');
          expect(getWarnings(newStats)).toMatchSnapshot('errors');
          expect(getErrors(newStats)).toMatchSnapshot('warnings');

          gzipSpy.mockRestore();

          resolve();
        });
      });
    });
  });
}
