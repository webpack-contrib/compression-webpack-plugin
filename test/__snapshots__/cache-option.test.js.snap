// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when applied with \`cache\` option matches snapshot for \`false\` value ({Boolean}): assets 1`] = `
Array [
  Array [
    "0.async.dbe59b7ea0e1a63e45d4.js",
    130,
  ],
  Array [
    "0.async.dbe59b7ea0e1a63e45d4.js.gz",
    117,
  ],
  Array [
    "5b1f36bc41ab31f5b801d48ba1d65781.png",
    78117,
  ],
  Array [
    "5b1f36bc41ab31f5b801d48ba1d65781.png.gz",
    73160,
  ],
  Array [
    "96621d3c37d96ad3bf792fcc848d912f.svg",
    672,
  ],
  Array [
    "96621d3c37d96ad3bf792fcc848d912f.svg.gz",
    393,
  ],
  Array [
    "js.8bd7079c9d63d0f6847e.js",
    10232,
  ],
  Array [
    "js.8bd7079c9d63d0f6847e.js.gz",
    2800,
  ],
]
`;

exports[`when applied with \`cache\` option matches snapshot for \`false\` value ({Boolean}): errors 1`] = `Array []`;

exports[`when applied with \`cache\` option matches snapshot for \`false\` value ({Boolean}): warnings 1`] = `Array []`;

exports[`when applied with \`cache\` option matches snapshot for \`other-cache-directory\` value ({String}): 0.async.dbe59b7ea0e1a63e45d4.js 1`] = `
Array [
  "0.async.dbe59b7ea0e1a63e45d4.js",
  "5c060e109cf6db498e14403ffb1625d2",
]
`;

exports[`when applied with \`cache\` option matches snapshot for \`other-cache-directory\` value ({String}): 5b1f36bc41ab31f5b801d48ba1d65781.png 1`] = `
Array [
  "5b1f36bc41ab31f5b801d48ba1d65781.png",
  "09a1a1112c577c2794359715edfcb5ac",
]
`;

exports[`when applied with \`cache\` option matches snapshot for \`other-cache-directory\` value ({String}): 96621d3c37d96ad3bf792fcc848d912f.svg 1`] = `
Array [
  "96621d3c37d96ad3bf792fcc848d912f.svg",
  "23fc1d3ac606d117e05a140e0de79806",
]
`;

exports[`when applied with \`cache\` option matches snapshot for \`other-cache-directory\` value ({String}): assets 1`] = `
Array [
  Array [
    "0.async.dbe59b7ea0e1a63e45d4.js",
    130,
  ],
  Array [
    "0.async.dbe59b7ea0e1a63e45d4.js.gz",
    117,
  ],
  Array [
    "5b1f36bc41ab31f5b801d48ba1d65781.png",
    78117,
  ],
  Array [
    "5b1f36bc41ab31f5b801d48ba1d65781.png.gz",
    73160,
  ],
  Array [
    "96621d3c37d96ad3bf792fcc848d912f.svg",
    672,
  ],
  Array [
    "96621d3c37d96ad3bf792fcc848d912f.svg.gz",
    393,
  ],
  Array [
    "js.8bd7079c9d63d0f6847e.js",
    10232,
  ],
  Array [
    "js.8bd7079c9d63d0f6847e.js.gz",
    2800,
  ],
]
`;

exports[`when applied with \`cache\` option matches snapshot for \`other-cache-directory\` value ({String}): assets 2`] = `
Array [
  Array [
    "0.async.dbe59b7ea0e1a63e45d4.js",
    130,
  ],
  Array [
    "0.async.dbe59b7ea0e1a63e45d4.js.gz",
    117,
  ],
  Array [
    "5b1f36bc41ab31f5b801d48ba1d65781.png",
    78117,
  ],
  Array [
    "5b1f36bc41ab31f5b801d48ba1d65781.png.gz",
    73160,
  ],
  Array [
    "96621d3c37d96ad3bf792fcc848d912f.svg",
    672,
  ],
  Array [
    "96621d3c37d96ad3bf792fcc848d912f.svg.gz",
    393,
  ],
  Array [
    "js.8bd7079c9d63d0f6847e.js",
    10232,
  ],
  Array [
    "js.8bd7079c9d63d0f6847e.js.gz",
    2800,
  ],
]
`;

exports[`when applied with \`cache\` option matches snapshot for \`other-cache-directory\` value ({String}): errors 1`] = `Array []`;

exports[`when applied with \`cache\` option matches snapshot for \`other-cache-directory\` value ({String}): errors 2`] = `Array []`;

exports[`when applied with \`cache\` option matches snapshot for \`other-cache-directory\` value ({String}): js.8bd7079c9d63d0f6847e.js 1`] = `
Array [
  "js.8bd7079c9d63d0f6847e.js",
  "9b0ef9dd9650dedb67c1a68fa7ab4c30",
]
`;

exports[`when applied with \`cache\` option matches snapshot for \`other-cache-directory\` value ({String}): warnings 1`] = `Array []`;

exports[`when applied with \`cache\` option matches snapshot for \`other-cache-directory\` value ({String}): warnings 2`] = `Array []`;

exports[`when applied with \`cache\` option matches snapshot for \`true\` value ({Boolean}): 0.async.dbe59b7ea0e1a63e45d4.js 1`] = `
Array [
  "0.async.dbe59b7ea0e1a63e45d4.js",
  "5c060e109cf6db498e14403ffb1625d2",
]
`;

exports[`when applied with \`cache\` option matches snapshot for \`true\` value ({Boolean}): 5b1f36bc41ab31f5b801d48ba1d65781.png 1`] = `
Array [
  "5b1f36bc41ab31f5b801d48ba1d65781.png",
  "09a1a1112c577c2794359715edfcb5ac",
]
`;

exports[`when applied with \`cache\` option matches snapshot for \`true\` value ({Boolean}): 96621d3c37d96ad3bf792fcc848d912f.svg 1`] = `
Array [
  "96621d3c37d96ad3bf792fcc848d912f.svg",
  "23fc1d3ac606d117e05a140e0de79806",
]
`;

exports[`when applied with \`cache\` option matches snapshot for \`true\` value ({Boolean}): assets 1`] = `
Array [
  Array [
    "0.async.dbe59b7ea0e1a63e45d4.js",
    130,
  ],
  Array [
    "0.async.dbe59b7ea0e1a63e45d4.js.gz",
    117,
  ],
  Array [
    "5b1f36bc41ab31f5b801d48ba1d65781.png",
    78117,
  ],
  Array [
    "5b1f36bc41ab31f5b801d48ba1d65781.png.gz",
    73160,
  ],
  Array [
    "96621d3c37d96ad3bf792fcc848d912f.svg",
    672,
  ],
  Array [
    "96621d3c37d96ad3bf792fcc848d912f.svg.gz",
    393,
  ],
  Array [
    "js.8bd7079c9d63d0f6847e.js",
    10232,
  ],
  Array [
    "js.8bd7079c9d63d0f6847e.js.gz",
    2800,
  ],
]
`;

exports[`when applied with \`cache\` option matches snapshot for \`true\` value ({Boolean}): assets 2`] = `
Array [
  Array [
    "0.async.dbe59b7ea0e1a63e45d4.js",
    130,
  ],
  Array [
    "0.async.dbe59b7ea0e1a63e45d4.js.gz",
    117,
  ],
  Array [
    "5b1f36bc41ab31f5b801d48ba1d65781.png",
    78117,
  ],
  Array [
    "5b1f36bc41ab31f5b801d48ba1d65781.png.gz",
    73160,
  ],
  Array [
    "96621d3c37d96ad3bf792fcc848d912f.svg",
    672,
  ],
  Array [
    "96621d3c37d96ad3bf792fcc848d912f.svg.gz",
    393,
  ],
  Array [
    "js.8bd7079c9d63d0f6847e.js",
    10232,
  ],
  Array [
    "js.8bd7079c9d63d0f6847e.js.gz",
    2800,
  ],
]
`;

exports[`when applied with \`cache\` option matches snapshot for \`true\` value ({Boolean}): errors 1`] = `Array []`;

exports[`when applied with \`cache\` option matches snapshot for \`true\` value ({Boolean}): errors 2`] = `Array []`;

exports[`when applied with \`cache\` option matches snapshot for \`true\` value ({Boolean}): js.8bd7079c9d63d0f6847e.js 1`] = `
Array [
  "js.8bd7079c9d63d0f6847e.js",
  "9b0ef9dd9650dedb67c1a68fa7ab4c30",
]
`;

exports[`when applied with \`cache\` option matches snapshot for \`true\` value ({Boolean}): warnings 1`] = `Array []`;

exports[`when applied with \`cache\` option matches snapshot for \`true\` value ({Boolean}): warnings 2`] = `Array []`;
